### PR TITLE
Add ability to use LLMs and embedding models from Ollama running locally.

### PR DESCRIPTION
**Pull request by damonsk**
Work done

- Specify Ollama address in settings.
- Implemented new Ollama provider, using OpenAI client which is compatible with Ollama.  This needs to override headers which are sent which breaks with Ollama (x-stainless-*)
- Added LLMs - llama3.1, incept5/llama3.1-claude
- Added Embeddings - nomic-embed-text, mxbai-embed-large
- Added required changes to drizzle

Tested on macOS.

### CODE REVIEW

File: src/utils/llm/manager.ts [hunk 1]
```diff
@@ -27,17 +28,28 @@ class LLMManager implements LLMManagerInterface {
...
+    apiKeys: { openai?: string; groq?: string; anthropic?: string },
+    ollamaBaseUrl?: string,
+  ) {
     this.openaiProvider = new OpenAIProvider(apiKeys.openai ?? '')
     this.groqProvider = new GroqProvider(apiKeys.groq ?? '')
     this.anthropicProvider = new AnthropicProvider(apiKeys.anthropic ?? '')
+    this.ollamaProvider = new OllamaAIOpenAIProvider(
+      apiKeys.openai ?? 'empty',
+      ollamaBaseUrl ?? '',
+    )
```
Comment (kevin-on) [inline src/utils/llm/manager.ts:43]: Could you explain why we're using the OpenAI API key with Ollama? As far as I understand, Ollama doesn't require an OpenAI API key to function.

Comment (damonsk) [reply src/utils/llm/manager.ts:43]: that's correct. During some research I saw a use case of hosted ollama with auth tokens but checking - it isn't part of the core functionality of ollama.  I will rework.

Comment (kevin-on) [reply src/utils/llm/manager.ts:43]: I see. Maybe we can add auth tokens later if needed. Thanks for your work.

Comment (damonsk) [reply src/utils/llm/manager.ts]: refactored this away in comment https://github.com/glowingjade/obsidian-smart-composer/pull/45/commits/043030b6acd437f53ac0ed5298ea6329c36c3b4a

Comment (glowingjade) [reply src/utils/llm/manager.ts]: Nice refactoring!

**Comment by glowingjade** [comment]
Nice work!

Can you fix migration file names to meaningful name? And also you have to update drizzle/meta/_journal.json afterwards.

Some thoughts
1. Each time we add an embedding model, we have to create a new table. While pgvector supports variable vector dimensions (https://github.com/pgvector/pgvector/issues/426), I'm not sure which approach is better - multiple tables vs. one table with variable vector dimensions. The former might be more performant? We can revisit this issue later though.

2. llama3.1:8b seems to have lower performance and struggles to understand our prompts properly. Smaller models tend to perform poorly when given too many instructions, so we should consider simplifying the prompts for smaller models. This is outside the scope of this PR though, so I'll add it to the backlog.

3. Since Ollama offers many models for users to choose from, it might be better to have a text input field rather than a dropdown in the settings page. However, this change needs some careful consideration, so I'll add it to the backlog instead of including it in this PR.

Everything else looks good. Once you update the migration file names, I'll merge this.

**Comment by damonsk** [comment]
Thanks @glowingjade, latest commit includes renames https://github.com/glowingjade/obsidian-smart-composer/pull/45/commits/8ca74f7e77b398465e289f0f7c2d1294c13c5ad2

I had noticed the models struggling with the prompts a litte and think your suggestions on allowing user to choose will help.

**Review by glowingjade — APPROVED** [review]

**Comment by glowingjade** [comment]
@damonsk Thank you for your contribution! We plan to release this feature in a few days.

**Pull request merged**